core/src/main/java/edu/wpi/grip/core/Pipeline.java

```diff
@@ -151,20 +249,24 @@ public void onSourceRemoved(SourceRemovedEvent event) {
         }
     }
 
-    public synchronized void addStep(int index, Step step) {
+    public void addStep(int index, Step step) {
```

Comment:
Spot the deadlock?
```
lock.lock();
try {
    eventBus.register(step);
    eventBus.post(new StepAddedEvent(step, index));
} finally {
    lock.unlock();
}
```
If you wrote that code, congratulations, you just deadlocked the application.
You should never post events from within a lock.

What I have done is write convenient helper methods for locking and unlocking the pipeline safely.
If we ever find that that core lock/unlock procedure is faulty we don't need to modify the same code in multiple places. This removes duplicate code.